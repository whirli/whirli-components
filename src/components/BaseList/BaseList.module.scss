@import '../../styles/Breakpoints.module.scss';

// default styles
.list {
    margin: 0;
    padding: 0;
    display: flex;
}

.direction-column {
    flex-direction: column;
}

.direction-row {
    flex-direction: row;
}

.style-disc {
    list-style: disc;
}

.style-none {
    list-style: none;
}
// breakpoint styles
$packageBreakpointStyles: ();