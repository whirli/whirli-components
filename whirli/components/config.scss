@import '@whirli-components/styles/Helpers.scss';

$breakpointPrefix: BREAKPOINT_;

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 672px,
    ),
    small_desk: (
        prefix: sd,
        min-width: 1056px
    ),
    large_desk: (
        prefix: ld,
        min-width: 1312px
    ),
    huge_desk: (
        prefix: hd,
        min-width: 1584px
    ),
);

:export {
    prefix: $breakpointPrefix;
    #{$breakpointPrefix}default: 0px;
    @each $name, $value in $breakpoints {
        #{$breakpointPrefix}#{$name}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}