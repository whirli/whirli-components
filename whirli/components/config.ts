import BaseAside from './BaseAside/BaseAside.config';
import BaseButton from './BaseButton/BaseButton.config';
import BaseColumn from './BaseColumn/BaseColumn.config';
import BaseContainer from './BaseContainer/BaseContainer.config';
import BaseGrid from './BaseGrid/BaseGrid.config';
import BaseImage from './BaseImage/BaseImage.config';
import BaseMain from './BaseMain/BaseMain.config';
import BaseSection from './BaseSection/BaseSection.config';
import BaseSpacer from './BaseSpacer/BaseSpacer.config';
import BaseText from './BaseText/BaseText.config';

export default {
  BaseAside,
  BaseButton,
  BaseColumn,
  BaseContainer,
  BaseGrid,
  BaseImage,
  BaseMain,
  BaseSection,
  BaseSpacer,
  BaseText,
};
