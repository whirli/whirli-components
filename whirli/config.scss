@import '@whirli-components/src/styles/Helpers.scss';

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 48rem,
    ),
    desktop: (
        prefix: d,
        min-width: 80rem,
    )
);

:export {
    @each $name, $value in $breakpoints {
        BREAKPOINT_#{to-upper-case(#{$name})}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}