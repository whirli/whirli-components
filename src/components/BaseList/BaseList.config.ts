// @ts-ignore
import { PropType } from 'vue';
import * as PropHelpers from '../../helpers/props';
import { ComponentStyles, ComponentProps } from '../../@types/components';

// Local config
import config from '@whirli-local/components/BaseList/BaseList.config';

// Package config
import { PropKeys, PropValues } from './BaseList.constants';
import type { PropStyle, PropTag } from './BaseList.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.LIST_STYLE]: {
    classes: {
      default: 'style-none',
      disc: 'style-disc',
      ...config.styles[PropKeys.LIST_STYLE]?.classes,
    } as Record<PropStyle, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.TAG]: {
    required: false,
    default: 'ul',
    type: String as PropType<PropTag>,
  },
  [PropKeys.LIST_STYLE]: {
    required: false,
    type: String as PropType<PropTag>,
    default: 'default',
    ...PropHelpers.stylePropValidation(ConfigStyles, PropKeys.LIST_STYLE),
  },
  ...config.props,
};
