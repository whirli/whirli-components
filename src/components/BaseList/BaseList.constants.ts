// Local config
import config from '@whirli-local/components/BaseList/BaseList.config';

// Helpers
import { getLocalPropValues } from '@whirli-components/helpers/props';

export const PropKeys: Record<string, string> = {
  TAG: 'tag',
  LIST_STYLE: 'listStyle',
  DIRECTION: 'direction',
};

const localValues = getLocalPropValues(PropKeys, config);

export const PropValues: Record<string, string[]> = {
  [PropKeys.TAG]: ['ul', 'ol'],
  [PropKeys.LIST_STYLE]: ['none', 'disc', ...localValues[PropKeys.LIST_STYLE]],
  [PropKeys.DIRECTION]: ['default', 'row', ...localValues[PropKeys.DIRECTION]],
};
