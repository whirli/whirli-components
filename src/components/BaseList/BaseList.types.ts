import { PropKeys, PropValues } from './BaseList.constants';

const propTag: string[] = PropValues[PropKeys.Tag];
const propStyle: string[] = PropValues[PropKeys.LIST_STYLE];

export type PropTag = typeof propTag[number];
export type PropStyle = typeof propStyle[number];
