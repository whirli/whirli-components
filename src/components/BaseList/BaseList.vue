<template>
  <component :is="props.tag" :class="classes"><slot /></component>
</template>

<script setup lang="ts">
// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseList/BaseList.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseList.config';

// Types
import { ComponentStyles as ComponentStylesInterface } from '@whirli-components/@types/components';
import { Props } from '@whirli-components/@types/props';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

// @ts-ignore
const props: Props = defineProps(ConfigProps);

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const classes = [styles.list, ...makeClasses(ComponentStyles, ConfigProps, props, styles)];
</script>
