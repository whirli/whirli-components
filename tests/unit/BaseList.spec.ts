import { mount } from '@vue/test-utils';

import { propReturnsClass } from '../helpers';

import BaseList from '@/components/BaseList/BaseList.vue';

import { PropKeys } from '@/components/BaseList/BaseList.constants';
import { ConfigStyles } from '@/components/BaseList/BaseList.config';

let wrapper: any;
const text = 'Hello world';

describe('BaseList.vue - Single classes', () => {
  beforeEach(() => {
    wrapper = mount(BaseList, {
      props: {
        tag: 'ul',
      },
      slots: {
        default: text,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct element', () => {
    expect(wrapper.find('ul').exists()).toBeTruthy();
  });

  it('renders the correct text', () => {
    expect(wrapper.text()).toBe(text);
  });

  it('is the correct classes', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.LIST_STYLE, 'style-none')).toBe(true);
  });
});
