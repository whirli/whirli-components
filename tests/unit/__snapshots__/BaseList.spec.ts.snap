// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseList.vue - Single classes renders the component 1`] = `
<ul
  class=""
>
  
  Hello world
  
</ul>
`;
